core/src/test/java/com/google/cloud/sql/core/RefresherTest.java

```diff
@@ -487,6 +487,65 @@ public void testCloudSqlRefreshesTokenIfExpired() throws Exception {
     assertThat(refreshCount.get()).isEqualTo(2);
   }
 
+  @Test
+  public void testGetConnectionInfo_throwsTerminalException_refreshOperationNotScheduled() {
+    ExampleData data = new ExampleData(Instant.now().plus(1, ChronoUnit.HOURS));
+    AtomicInteger refreshCount = new AtomicInteger();
+
+    Refresher r =
+        new Refresher(
+            "RefresherTest.testGetConnectionInfo_throwsTerminalException_refreshOperationNotScheduled",
+            executorService,
+            () -> {
+              int c = refreshCount.get();
+              ExampleData refreshResult = data;
+              switch (c) {
+                case 0:
+                  // refresh 0 should throw an exception
+                  refreshCount.incrementAndGet();
+                  throw new TerminalException("Not authorized");
+              }
+              // refresh 2 and on should return data immediately
+              refreshCount.incrementAndGet();
+              return Futures.immediateFuture(refreshResult);
+            },
+            rateLimiter);
+
+    assertThrows(TerminalException.class, () -> r.getConnectionInfo(TEST_TIMEOUT_MS));
+    assertThat(refreshCount.get()).isEqualTo(1);
```

Comment:
Does this assertion fail before the changes above? Is there a race condition here where this might pass before the refresher's executor would run the next task?

Reply:
A comment explaining why this assertion works with the time ut above might help.

Reply:
The assertion on line 514 fails before implementing the changes above because no exception is raised by `getConnectionInfo` when the refresher's executor runs the next task.

Added a comment.